src/System.Private.CoreLib/shared/System/Double.cs

```diff
@@ -90,8 +90,12 @@ public static unsafe bool IsInfinity(double d)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static unsafe bool IsNaN(double d)
         {
-            long bits = BitConverter.DoubleToInt64Bits(d);
-            return (bits & 0x7FFFFFFFFFFFFFFF) > 0x7FF0000000000000;
+            // A NaN will never equal itself so this is an
+            // easy and efficient way to check for NaN.
+
+            #pragma warning disable CS1718
+            return d != d;
+            #pragma warning restore CS1718
```

Comment:
Just wondering: if we add a kind of a `-ffast-math` mode to CLR what will `double.IsNaN(double.NaN)` return?

Reply:
It wouldn't apply to every method automatically, it would need to be opt in exactly for cases like this.

Reply:
Well, this code is shared with mono and we do have ffast-math and this check becomes `return false` even for `Double.IsNaN` but it's probably expected when you use ffast-math (which implies no nans).
Also since it's shared I am not sure it won't break IEEE754 compatibility for non-llvm mono (checking) and CoreRT

Reply:
Does Mono have an attribute that can be used to skip fast math for certain methods?

As an aside, it seems very undesirable to impact methods that haven't opted into fast math behavior, especially when they may be relying on semantics like this.

> IEEE754 for non-LLVM mono

If this breaks regular mono, that is an IEEE 754 compliance bug that Mono (or other runtimes, like CoreRT) should likely fix 😄

Reply:
@tannergooding yeah I've edited my message, `!=` operator is OK in mono 🙂 

> Does Mono have an attribute that can be used to skip fast math for certain methods?
Not yet but it shouldn't be difficult to add (not sure what will(should) happen with some method will inline my `[no-fast-math]` method).

Reply:
In general, I think that things like `fast math` need to be opt in; and there needs to be a way to explicitly say methods can't be "fast mathed".

WinForms, WPF, ML.NET, and others frequently use `NaN` to represent "invalid" data and so they need to preserve `NaN` semantics in a good number of locations and just automatically opting into `fast math` semantics will break them.

This is, naturally, all stuff we'll need to consider if we want a "fast math" feature in CoreCLR itself.